reframe/core/meta.py

```diff
@@ -417,19 +433,8 @@ class attribute. This behavior does not apply when the assigned value
         is not allowed. This would break the parameter space internals.
         '''
 
-        # Set the value of a variable (except when the value is a descriptor).
         try:
-            var_space = super().__getattribute__('_rfm_var_space')
-            if name in var_space:
-                if not hasattr(value, '__get__'):
-                    var_space[name].define(value)
-                    return
-                elif not var_space[name].field is value:
-                    desc = '.'.join([cls.__qualname__, name])
-                    raise ReframeSyntaxError(
-                        f'cannot override variable descriptor {desc!r}'
-                    )
-
+            cls.setvar(name, value)
```

Comment:
If no `AttributeError` is raised here, this function has to return because the variable's value has been set and there's nothing else to do here.
```suggestion
            cls.setvar(name, value)
            return
```

Reply:
This still works fine because all the namespace clashing checks happen during class creation, and this is after that. So even if without this return here, the default `__setattr__` below is called, this attribute gets silently replaced by the variable's descriptor during the first instantiation.

Reply:
Actually, without this return here, the `__getattribute__` above might not retrieve the right value, since the variables only take precedence over regular attributes in the class `__dict__` after the first class instantiation. In any case, this value and the variable's default value would always be the same.

Reply:
You are right. Factoring out the the code to `setvar()` is not equivalent to the code before, because previously you were returning from the `__setattr__()` and now I'm only returning from `setvar()` 😬

Reply:
It needs another little twist apart from the `return`. We need to handle the case when the class variables are converted to their internal descriptors at injection time. Previously, this was falling through to the final `super().__setattr_(name, value)`, but by putting just the `return` here, we cut that fall through, causing everything to fail. We need an explicit `else` in the `setvar()` where we're checking whether we're actually injecting the descriptor or not and call the superclass' `__setattr__` explicitly.

Reply:
Ah true! I'd suggest that we do this a little different then. How about making `setvar()` return a `bool` indicating whether the variable was set or not? With this, the bit in `__setattr__()` would just be
```python
if cls.setvar(name, value):
    return
```
`setvar()` could also return `False` when injecting the descriptor, so that it would fall through to the base `__setattr__`.

Reply:
I've also commented properly the `setvar()` now for future reference.

Reply:
Maybe we should add in there "change this at your own risk" 😂